Improve LSP1-UniversalReceiver Standard (#144)
lukso-network/LIPs: LSPs/LSP-1-UniversalReceiver.md

## Specification

**LSP1-UniversalReceiver** interface id according to [ERC165]: `0x6bb56a14`.

Smart contracts implementing the LSP1UniversalReceiver standard MUST implement the [ERC165] `supportsInterface(..)` function and MUST support ERC165 and LSP1 interface ids.

Every contract that complies with the LSP1-UniversalReceiver standard MUST implement:

### Methods

#### universalReceiver

```solidity

> function universalReceiver(bytes32 typeId, bytes memory data) external payable returns (bytes memory)

Reviewer: Very good. `external` is definitely the right visibility for this standard. Since it is used for external actors to inform.
And if people want to "open up" the function to the contract internally, they can override to `public`.

I agree on putting this 👍 

For references:

This compiles ✅ 

<img width="427" alt="image" src="https://user-images.githubusercontent.com/31145285/204538956-1539c4f6-f62e-46ee-99a2-7aa1fce1353f.png">

This does not compile ❌ 

<img width="416" alt="image" src="https://user-images.githubusercontent.com/31145285/204539118-36a425d5-6911-4ad9-adc4-9ccd7b2933d3.png">

## UniversalReceiverDelegate

> The ability to react to upcoming actions with a logic hardcoded within the `universalReceiver(..)` function comes with a few limitations. One is the possibility of missing on reacting on a few cases, leading to new deployment and migrating to a new contract that supports responding to these missed cases. This can be problematic, primarily if third parties are related to the contract.

Reviewer 2: This should be in the motivation section, not in the spec part

Reviewer 2: Please look at the default format of an EIP document,. the motivation section is not within the spec. Its is own section on the top.

https://github.com/lukso-network/LIPs/blob/main/lip-X.md

Reviewer 2: I see now you structured the document in two specifications... let me do edits myself...

> The ability to react to upcoming actions with a logic hardcoded within the `universalReceiver(..)` function is good but comes with a few limitations, one of them is the possibility to miss on reacting on a few cases which leads to new deployment and migrating to a new contract that supports reacting on the missed cases. This can be really bad specially if there are third parties that are related to the contract.

Reviewer: ```suggestion
The ability to react to upcoming actions with a logic hardcoded within the `universalReceiver(..)` function comes with a few limitations. One is the possibility of missing on reacting on a few cases, leading to new deployment and migrating to a new contract that supports responding to these missed cases. This can be problematic, primarily if third parties are related to the contract.
```

> Another limitation is sticking to one behavior, sometimes there are cases, specially with smart contract-based accounts, where you want to change the way you react on different assets after some time, e.g, you may want to reject all the assets that you are being informed about, and in a certain point of time, you want to accept specific tokens that are verified by a third party registry.  

Reviewer: ```suggestion
Another limitation is sticking to one behaviour without the ability to make it change. In some cases, like with smart contract-based accounts, there might be a need to adjust how the contract reacts to certain received assets over time. For instance, a user might want to reject any assets in the first place but decide at some point to accept specific tokens that a third-party registry has verified.
```

> The **UniversalReceiverDelegate** is an optional extension that could be used with the `universalReceiver(..)` function that will allow the function to delegate its universalReceiver functionality to an external contract that can be customized to react in different ways. 

Reviewer: ```suggestion
The **UniversalReceiverDelegate** is an optional extension. It allows the `universalReceiver(..)` function to delegate its functionality to an external contract that can be customized to react differently based on the `typeId` and the `data` received.
```

> The `universalReceiver(..)` function forwards the call being received to the `universalReceiver(..)` on the **UniversalReceiverDelegate** contract and can append the calldata with extra 52 bytes as follows:

Reviewer: Below you say _"can append the calldata ..._

So appending extra infos is optional? Or mandatory as per the standard?

```suggestion
The `universalReceiver(..)` function forwards the call being received to the `universalReceiver(..)` on the **UniversalReceiverDelegate** contract and can append the calldata with 52 extra bytes as follows:
```
